crates/ast/src/grammar.lalrpop

```diff
@@ -0,0 +1,227 @@
+use alloy_dyn_abi::DynSolType;
+use alloy_primitives::{hex::FromHex, Bytes, U256};
+use evm_glue::opcodes::Opcode;
+use lalrpop_util::ParseError;
+use std::str::FromStr;
+use crate as ast;
+use crate::parser::{u256_as_push, u256_as_push_data};
+
+grammar;
+
+extern {
+    type Error = ast::Error;
+}
+
+pub Root: ast::Root<'input> = {
+    <def:Definition*> => ast::Root(def.into_iter().collect()),
+};
+
+pub Definition: ast::HuffDefinition<'input> = {
+    "#define" <m:Macro> => m,
+    "#define" <c:Constant> => c,
+    "#define" <t:Table> => t,
+    "#define" <f:SolFunction> => f,
+    "#define" <e:SolEvent> => e,
+    "#define" <e:SolError> => e,
+};
+
+pub Macro: ast::HuffDefinition<'input> = {
+    "macro" <name:IDENT> <args:MacroArgs> "=" "takes" "(" <takes:DEC> ")" "returns" "(" <returns:DEC> ")" "{" <insts:Instruction+> "}" => {
```

Comment:
Not familiar with LALRPOP but does `+` mean "1 or more"? I think macros can be empty.

Also in the original Huff `takes` & `returns` are optional e.g. the following is allowed:

```
#define macro WOW() = {
    
}
```

Maybe we should allow that too for the sake of backwards compatibility?